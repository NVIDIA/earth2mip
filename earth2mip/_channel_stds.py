# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

channel_stds = {
    "u10m": 5.582091331481934,
    "v10m": 4.789056301116943,
    "u100m": 7.004323482513428,
    "v100m": 6.123928546905518,
    "t2m": 21.191524505615234,
    "sp": 9595.267578125,
    "msl": 1344.2525634765625,
    "tcwv": 16.540952682495117,
    "u1": 38.58013916015625,
    "u2": 34.72814178466797,
    "u3": 32.2275390625,
    "u5": 29.21213722229004,
    "u7": 27.27058219909668,
    "u10": 25.552366256713867,
    "u20": 21.857524871826172,
    "u30": 19.176223754882812,
    "u50": 15.317777633666992,
    "u70": 13.463854789733887,
    "u100": 13.648713111877441,
    "u125": 14.856287002563477,
    "u150": 16.07061767578125,
    "u175": 17.02838897705078,
    "u200": 17.721553802490234,
    "u225": 18.051687240600586,
    "u250": 18.01706886291504,
    "u300": 17.13730239868164,
    "u350": 15.746345520019531,
    "u400": 14.337974548339844,
    "u450": 13.082659721374512,
    "u500": 12.004372596740723,
    "u550": 11.114582061767578,
    "u600": 10.387426376342773,
    "u650": 9.774086952209473,
    "u700": 9.236701965332031,
    "u750": 8.800454139709473,
    "u775": 8.619154930114746,
    "u800": 8.46338176727295,
    "u825": 8.338661193847656,
    "u850": 8.247283935546875,
    "u875": 8.179678916931152,
    "u900": 8.111303329467773,
    "u925": 7.978816032409668,
    "u950": 7.672853946685791,
    "u975": 7.052810192108154,
    "u1000": 6.180253028869629,
    "v1": 15.007347106933594,
    "v2": 14.02863883972168,
    "v3": 13.400466918945312,
    "v5": 12.586094856262207,
    "v7": 12.008319854736328,
    "v10": 11.252314567565918,
    "v20": 9.49392318725586,
    "v30": 8.383625984191895,
    "v50": 7.201364517211914,
    "v70": 6.877016067504883,
    "v100": 7.557221412658691,
    "v125": 8.542673110961914,
    "v150": 9.602500915527344,
    "v175": 10.747245788574219,
    "v200": 11.92225456237793,
    "v225": 12.88244342803955,
    "v250": 13.436037063598633,
    "v300": 13.366893768310547,
    "v350": 12.399846076965332,
    "v400": 11.22365951538086,
    "v450": 10.125510215759277,
    "v500": 9.184946060180664,
    "v550": 8.425360679626465,
    "v600": 7.8204240798950195,
    "v650": 7.334446430206299,
    "v700": 6.889240264892578,
    "v750": 6.559966087341309,
    "v775": 6.439513206481934,
    "v800": 6.349121570587158,
    "v825": 6.294491291046143,
    "v850": 6.284781455993652,
    "v875": 6.329617500305176,
    "v900": 6.4216156005859375,
    "v925": 6.501189708709717,
    "v950": 6.4423980712890625,
    "v975": 6.05946159362793,
    "v1000": 5.338335037231445,
    "w1": 0.0009700588998384774,
    "w2": 0.0017569169867783785,
    "w3": 0.002291186945512891,
    "w5": 0.0030673006549477577,
    "w7": 0.0038343495689332485,
    "w10": 0.004771752282977104,
    "w20": 0.007248147390782833,
    "w30": 0.0092012332752347,
    "w50": 0.012652132660150528,
    "w70": 0.016436472535133362,
    "w100": 0.02630731090903282,
    "w125": 0.038530394434928894,
    "w150": 0.05268237739801407,
    "w175": 0.06788258254528046,
    "w200": 0.08365459740161896,
    "w225": 0.09980383515357971,
    "w250": 0.11605729162693024,
    "w300": 0.1480696201324463,
    "w350": 0.1759636253118515,
    "w400": 0.19732540845870972,
    "w450": 0.2122214138507843,
    "w500": 0.22131618857383728,
    "w550": 0.22764532268047333,
    "w600": 0.23328272998332977,
    "w650": 0.23892800509929657,
    "w700": 0.24304194748401642,
    "w750": 0.24660667777061462,
    "w775": 0.24764472246170044,
    "w800": 0.24739933013916016,
    "w825": 0.24559083580970764,
    "w850": 0.24174177646636963,
    "w875": 0.23500598967075348,
    "w900": 0.223381906747818,
    "w925": 0.2050851732492447,
    "w950": 0.1813991665840149,
    "w975": 0.1576462835073471,
    "w1000": 0.14332318305969238,
    "z1": 17160.59375,
    "z2": 15051.7412109375,
    "z3": 13678.45703125,
    "z5": 12030.5791015625,
    "z7": 11000.9716796875,
    "z10": 9939.580078125,
    "z20": 7970.08447265625,
    "z30": 6934.6259765625,
    "z50": 5881.4306640625,
    "z70": 5497.146484375,
    "z100": 5545.66748046875,
    "z125": 5721.24609375,
    "z150": 5838.37841796875,
    "z175": 5871.626953125,
    "z200": 5826.142578125,
    "z225": 5712.2578125,
    "z250": 5541.61279296875,
    "z300": 5096.86474609375,
    "z350": 4614.15478515625,
    "z400": 4156.310546875,
    "z450": 3738.7666015625,
    "z500": 3361.076171875,
    "z550": 3018.40625,
    "z600": 2704.42138671875,
    "z650": 2413.234619140625,
    "z700": 2143.913818359375,
    "z750": 1898.7406005859375,
    "z775": 1784.596435546875,
    "z800": 1676.227783203125,
    "z825": 1573.9453125,
    "z850": 1478.46435546875,
    "z875": 1390.4481201171875,
    "z900": 1310.4091796875,
    "z925": 1238.84423828125,
    "z950": 1176.4459228515625,
    "z975": 1124.12060546875,
    "z1000": 1082.5850830078125,
    "t1": 12.217451095581055,
    "t2": 15.37381362915039,
    "t3": 15.199849128723145,
    "t5": 14.100061416625977,
    "t7": 13.47716236114502,
    "t10": 12.883852005004883,
    "t20": 11.544325828552246,
    "t30": 10.671643257141113,
    "t50": 10.337119102478027,
    "t70": 11.590261459350586,
    "t100": 12.401776313781738,
    "t125": 10.661434173583984,
    "t150": 8.748968124389648,
    "t175": 7.472810745239258,
    "t200": 7.142058849334717,
    "t225": 7.623969554901123,
    "t250": 8.593498229980469,
    "t300": 10.757156372070312,
    "t350": 12.078363418579102,
    "t400": 12.678500175476074,
    "t450": 12.939180374145508,
    "t500": 13.049148559570312,
    "t550": 13.15401840209961,
    "t600": 13.449394226074219,
    "t650": 14.161675453186035,
    "t700": 14.810254096984863,
    "t750": 15.129765510559082,
    "t775": 15.2506685256958,
    "t800": 15.36613655090332,
    "t825": 15.471822738647461,
    "t850": 15.559452056884766,
    "t875": 15.650764465332031,
    "t900": 15.786845207214355,
    "t925": 15.981817245483398,
    "t950": 16.2429141998291,
    "t975": 16.620704650878906,
    "t1000": 16.995594024658203,
    "r1": 0.01644953526556492,
    "r2": 0.045997656881809235,
    "r3": 0.0963272675871849,
    "r5": 0.38980916142463684,
    "r7": 1.0760670900344849,
    "r10": 3.1369779109954834,
    "r20": 13.323467254638672,
    "r30": 15.115065574645996,
    "r50": 15.014254570007324,
    "r70": 18.140913009643555,
    "r100": 33.699092864990234,
    "r125": 32.9305305480957,
    "r150": 32.013851165771484,
    "r175": 32.755245208740234,
    "r200": 34.16655731201172,
    "r225": 35.020626068115234,
    "r250": 34.8388671875,
    "r300": 34.21603775024414,
    "r350": 34.3671989440918,
    "r400": 34.230587005615234,
    "r450": 33.944515228271484,
    "r500": 33.60013198852539,
    "r550": 33.3435173034668,
    "r600": 32.77552795410156,
    "r650": 32.15131759643555,
    "r700": 31.51679229736328,
    "r750": 30.690650939941406,
    "r775": 30.05921745300293,
    "r800": 29.193891525268555,
    "r825": 28.011188507080078,
    "r850": 26.4312801361084,
    "r875": 24.70786476135254,
    "r900": 23.095003128051758,
    "r925": 21.721576690673828,
    "r950": 20.58991241455078,
    "r975": 19.374773025512695,
    "r1000": 18.462560653686523,
    "2d": 20.632549285888672,
    "sst": 136.16073608398438,
    "skt": 22.203195571899414,
    "stl1": 21.357025146484375,
    "swvl1": 0.14037616550922394,
}
