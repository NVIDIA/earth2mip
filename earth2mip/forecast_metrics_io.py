# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Routines for reading and writing forecast metrics to a directory of csv files.

The csv files contain the records::

    initial_time_iso, lead_time_hours, channel, metric, value
    2022-01-01T00:00:00,24,t2m,rmse,25.6

"""
import csv
import datetime
import os
from typing import IO

import pandas as pd


def read_metrics(directory: str) -> pd.Series:
    """
    Reads all csv files in the given directory and returns a pandas Series
    containing all the metric values.
    """

    dfs = []
    # Loop through all csv files in the directory
    csv_files = [
        filename for filename in os.listdir(directory) if filename.endswith(".csv")
    ]

    if len(csv_files) == 0:
        return pd.Series()

    for filename in csv_files:
        filepath = os.path.join(directory, filename)
        # Read the csv file into a pandas DataFrame
        df = pd.read_csv(
            filepath,
            header=None,
            names=[
                "initial_time_iso",
                "lead_time_hours",
                "channel",
                "metric",
                "value",
            ],
        )
        out = pd.DataFrame()
        out["initial_time"] = pd.to_datetime(df["initial_time_iso"])
        out["lead_time"] = pd.to_timedelta(df["lead_time_hours"], unit="h")
        out["channel"] = df["channel"]
        out["metric"] = df["metric"]
        out["value"] = df["value"]
        dfs.append(out)

    df = pd.concat(dfs, axis=0)
    df.set_index(["initial_time", "lead_time", "channel", "metric"], inplace=True)
    return df["value"]


def write_metric(
    f: IO[str],
    initial_time: datetime.datetime,
    lead_time: datetime.timedelta,
    channel: str,
    metric: str,
    value: float,
) -> None:
    """
    Writes a single metric value to the given file object in csv format.
    """
    # Write the metric value to the file object in csv format
    writer = csv.writer(f)
    writer.writerow(
        [
            initial_time.isoformat(),
            lead_time.total_seconds() // 3600,
            channel,
            metric,
            value,
        ]
    )
