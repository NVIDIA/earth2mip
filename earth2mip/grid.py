# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import dataclasses
from typing import List

import numpy as np

from earth2mip import schema


@dataclasses.dataclass(frozen=True)
class LatLonGrid:

    lat: List[float]
    lon: List[float]

    @property
    def shape(self):
        return (len(self.lat), len(self.lon))


def equiangular_lat_lon_grid(
    nlat: int, nlon: int, includes_south_pole: bool = True
) -> LatLonGrid:
    """A regular lat-lon grid

    Lat is ordered from 90 to -90. Includes -90 and only if if includes_south_pole is True.
    Lon is ordered from 0 to 360. includes 0, but not 360.

    """  # noqa
    lat = np.linspace(90, -90, nlat, endpoint=includes_south_pole)
    lon = np.linspace(0, 360, nlon, endpoint=False)
    return LatLonGrid(lat.tolist(), lon.tolist())


def from_enum(grid_enum: schema.Grid) -> LatLonGrid:
    if grid_enum == schema.Grid.grid_720x1440:
        return equiangular_lat_lon_grid(720, 1440, includes_south_pole=False)
    elif grid_enum == schema.Grid.grid_721x1440:
        return equiangular_lat_lon_grid(721, 1440)
    elif grid_enum == schema.Grid.s2s_challenge:
        return equiangular_lat_lon_grid(181, 360)
    else:
        raise ValueError(f"Unknown grid {grid_enum}")
