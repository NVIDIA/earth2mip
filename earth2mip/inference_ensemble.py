# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import json
import logging
import os
import sys
from datetime import datetime
from typing import Any, Optional

import cftime
import numpy as np
import torch
import tqdm
import xarray
from modulus.distributed.manager import DistributedManager
from netCDF4 import Dataset as DS

import earth2mip.grid

__all__ = ["run_inference", "run_basic_inference"]

# need to import initial conditions first to avoid unfortunate
# GLIBC version conflict when importing xarray. There are some unfortunate
# issues with the environment.
from earth2mip import initial_conditions, regrid, time_loop
from earth2mip._channel_stds import channel_stds
from earth2mip.ensemble_utils import (
    generate_bred_vector,
    generate_noise_correlated,
    generate_noise_grf,
)
from earth2mip.netcdf import initialize_netcdf, update_netcdf
from earth2mip.networks import get_model
from earth2mip.schema import EnsembleRun, PerturbationStrategy
from earth2mip.time_loop import TimeLoop

logger = logging.getLogger("inference")


def get_checkpoint_path(rank, batch_id, path):
    directory = os.path.join(path, f"{rank}")
    filename = f"{batch_id}.pth"
    return os.path.join(directory, filename)


def save_restart(restart, rank, batch_id, path):
    path = get_checkpoint_path(rank, batch_id, path)
    os.makedirs(os.path.dirname(path), exist_ok=True)
    logger.info(f"Saving restart file to {path}.")
    torch.save(restart, path)


def run_ensembles(
    *,
    n_steps: int,
    weather_event,
    model: TimeLoop,
    perturb,
    x,
    nc,
    domains,
    n_ensemble: int,
    batch_size: int,
    rank: int,
    output_frequency: int,
    output_grid: Optional[earth2mip.grid.LatLonGrid],
    date_obj: datetime,
    restart_frequency: Optional[int],
    output_path: str,
    restart_initial_directory: str = "",
    progress: bool = True,
):
    if not output_grid:
        output_grid = model.grid

    regridder = regrid.get_regridder(model.grid, output_grid).to(model.device)

    diagnostics = initialize_netcdf(nc, domains, output_grid, n_ensemble, model.device)
    initial_time = date_obj
    time_units = initial_time.strftime("hours since %Y-%m-%d %H:%M:%S")
    nc["time"].units = time_units
    nc["time"].calendar = "standard"

    for batch_id in range(0, n_ensemble, batch_size):
        logger.info(f"ensemble members {batch_id+1}-{batch_id+batch_size}/{n_ensemble}")
        batch_size = min(batch_size, n_ensemble - batch_id)

        x = x.repeat(batch_size, 1, 1, 1, 1)
        x = perturb(x, rank, batch_id, model.device)
        # restart_dir = weather_event.properties.restart

        # TODO: figure out if needed
        # if restart_dir:
        #     path = get_checkpoint_path(rank, batch_id, restart_dir)
        #     # TODO use logger
        #     logger.info(f"Loading from restart from {path}")
        #     kwargs = torch.load(path)
        # else:
        #     kwargs = dict(
        #         x=x,
        #         normalize=False,
        #         time=time,
        #     )

        iterator = model(initial_time, x)

        # Check if stdout is connected to a terminal
        if sys.stderr.isatty() and progress:
            iterator = tqdm.tqdm(iterator, total=n_steps)

        time_count = -1

        # for time, data, restart in iterator:

        for k, (time, data, _) in enumerate(iterator):
            # if restart_frequency and k % restart_frequency == 0:
            #     save_restart(
            #         restart,
            #         rank,
            #         batch_id,
            #         path=os.path.join(output_path, "restart", time.isoformat()),
            #     )

            # Saving the output
            if output_frequency and k % output_frequency == 0:
                time_count += 1
                logger.debug(f"Saving data at step {k} of {n_steps}.")
                nc["time"][time_count] = cftime.date2num(time, nc["time"].units)
                update_netcdf(
                    regridder(data),
                    diagnostics,
                    domains,
                    batch_id,
                    time_count,
                    model.grid,
                    model.out_channel_names,
                )

            if k == n_steps:
                break

        # if restart_frequency is not None:
        #     save_restart(
        #         restart,
        #         rank,
        #         batch_id,
        #         path=os.path.join(output_path, "restart", "end"),
        #     )


def main(config=None):
    logging.basicConfig(level=logging.INFO)

    if config is None:
        parser = argparse.ArgumentParser()
        parser.add_argument("config")
        parser.add_argument("--weather_model", default=None)
        args = parser.parse_args()
        config = args.config

    # If config is a file
    if os.path.exists(config):
        config: EnsembleRun = EnsembleRun.parse_file(config)
    # If string, assume JSON string
    elif isinstance(config, str):
        config: EnsembleRun = EnsembleRun.parse_obj(json.loads(config))
    # Otherwise assume parsable obj
    else:
        raise ValueError(
            f"Passed config parameter {config} should be valid file or JSON string"
        )

    # if args and args.weather_model:
    #     config.weather_model = args.weather_model

    # Set up parallel
    DistributedManager.initialize()
    device = DistributedManager().device
    group = torch.distributed.group.WORLD

    logging.info(f"Earth-2 MIP config loaded {config}")
    logging.info(f"Loading model onto device {device}")
    model = get_model(config.weather_model, device=device)
    logging.info("Constructing initializer data source")
    perturb = get_initializer(
        model,
        config,
    )
    logging.info("Running inference")
    run_inference(model, config, perturb, group)


def get_initializer(
    model,
    config,
):
    def perturb(x, rank, batch_id, device):
        shape = x.shape
        if config.perturbation_strategy == PerturbationStrategy.gaussian:
            noise = config.noise_amplitude * torch.normal(
                torch.zeros(shape), torch.ones(shape)
            ).to(device)
        elif config.perturbation_strategy == PerturbationStrategy.correlated:
            noise = generate_noise_correlated(
                shape,
                reddening=config.noise_reddening,
                device=device,
                noise_amplitude=config.noise_amplitude,
            )
        elif config.perturbation_strategy == PerturbationStrategy.spherical_grf:
            noise = generate_noise_grf(
                shape,
                model.grid,
                sigma=config.grf_noise_sigma,
                alpha=config.grf_noise_alpha,
                tau=config.grf_noise_tau,
                device=device,
            )
        elif config.perturbation_strategy == PerturbationStrategy.bred_vector:
            noise = generate_bred_vector(
                x,
                model,
                config.noise_amplitude,
                time=config.weather_event.properties.start_time,
            )
        elif config.perturbation_strategy == PerturbationStrategy.none:
            return x
        if rank == 0 and batch_id == 0:  # first ens-member is deterministic
            noise[0, :, :, :, :] = 0

        # When field is not in known normalization dictionary set scale to 0
        scale = []
        for i, channel in enumerate(model.in_channel_names):
            if channel in channel_stds:
                scale.append(channel_stds[channel])
            else:
                scale.append(0)
        scale = torch.tensor(scale, device=x.device)

        if config.perturbation_channels is None:
            x += noise * scale[:, None, None]
        else:
            channel_list = model.in_channel_names
            indices = torch.tensor(
                [
                    channel_list.index(channel)
                    for channel in config.perturbation_channels
                    if channel in channel_list
                ]
            )
            x[:, :, indices, :, :] += (
                noise[:, :, indices, :, :] * scale[indices, None, None]
            )
        return x

    return perturb


def run_basic_inference(
    model: time_loop.TimeLoop,
    n: int,
    data_source: Any,
    time: datetime,
):
    """Run a basic inference"""

    x = initial_conditions.get_initial_condition_for_model(model, data_source, time)

    arrays = []
    times = []
    for k, (time, data, _) in enumerate(model(time, x)):
        arrays.append(data.cpu().numpy())
        times.append(time)
        if k == n:
            break

    stacked = np.stack(arrays)
    coords = dict(lat=model.grid.lat, lon=model.grid.lon)
    coords["channel"] = model.out_channel_names
    coords["time"] = times
    return xarray.DataArray(
        stacked, dims=["time", "history", "channel", "lat", "lon"], coords=coords
    )


def run_inference(
    model: TimeLoop,
    config: EnsembleRun,
    perturb: Any = None,
    group: Any = None,
    progress: bool = True,
    # TODO add type hints
    data_source: Any = None,
):
    """Run an ensemble inference for a given config and a perturb function

    Args:
        group: the torch distributed group to use for the calculation
        progress: if True use tqdm to show a progress bar
        data_source: a Mapping object indexed by datetime and returning an
            xarray.Dataset object.
    """
    if not perturb:
        perturb = get_initializer(model, config)

    if not group and torch.distributed.is_initialized():
        group = torch.distributed.group.WORLD

    weather_event = config.get_weather_event()

    if not data_source:
        data_source = initial_conditions.get_data_source(
            model.in_channel_names,
            initial_condition_source=weather_event.properties.initial_condition_source,
            netcdf=weather_event.properties.netcdf,
        )

    date_obj = weather_event.properties.start_time
    x = initial_conditions.get_initial_condition_for_model(model, data_source, date_obj)

    dist = DistributedManager()
    n_ensemble_global = config.ensemble_members
    n_ensemble = n_ensemble_global // dist.world_size
    if n_ensemble == 0:
        logger.warning("World size is larger than global number of ensembles.")
        n_ensemble = n_ensemble_global

    # Set random seed
    seed = config.seed
    torch.manual_seed(seed + dist.rank)
    np.random.seed(seed + dist.rank)

    if config.output_dir:
        date_str = "{:%Y_%m_%d_%H_%M_%S}".format(date_obj)
        name = weather_event.properties.name
        output_path = (
            f"{config.output_dir}/"
            f"Output.{config.weather_model}."
            f"{name}.{date_str}"
        )
    else:
        output_path = config.output_path

    if not os.path.exists(output_path):
        # Avoid race condition across ranks
        os.makedirs(output_path, exist_ok=True)

    if dist.rank == 0:
        # Only rank 0 copies config files over
        config_path = os.path.join(output_path, "config.json")
        with open(config_path, "w") as f:
            f.write(config.json())

    group_rank = torch.distributed.get_group_rank(group, dist.rank)
    output_file_path = os.path.join(output_path, f"ensemble_out_{group_rank}.nc")

    with DS(output_file_path, "w", format="NETCDF4") as nc:
        # assign global attributes
        nc.model = config.weather_model
        nc.config = config.json()
        nc.weather_event = weather_event.json()
        nc.date_created = datetime.now().isoformat()
        nc.history = " ".join(sys.argv)
        nc.institution = "NVIDIA"
        nc.Conventions = "CF-1.10"

        run_ensembles(
            weather_event=weather_event,
            model=model,
            perturb=perturb,
            nc=nc,
            domains=weather_event.domains,
            x=x,
            n_ensemble=n_ensemble,
            n_steps=config.simulation_length,
            output_frequency=config.output_frequency,
            batch_size=config.ensemble_batch_size,
            rank=dist.rank,
            date_obj=date_obj,
            restart_frequency=config.restart_frequency,
            output_path=output_path,
            output_grid=(
                earth2mip.grid.from_enum(config.output_grid)
                if config.output_grid
                else None
            ),
            progress=progress,
        )
    if torch.distributed.is_initialized():
        torch.distributed.barrier(group)

    logger.info(f"Ensemble forecast finished, saved to: {output_file_path}")


if __name__ == "__main__":
    main()
