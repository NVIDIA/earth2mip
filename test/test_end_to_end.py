# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
# SPDX-FileCopyrightText: All rights reserved.
# SPDX-License-Identifier: Apache-2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import datetime
import hashlib
import pathlib
import subprocess
from test.initial_conditions.test_hdf5 import create_hdf5

import numpy as np
import pytest
import torch
import xarray

import earth2mip.forecast_metrics_io
import earth2mip.grid
import earth2mip.networks.dlwp
from earth2mip import (
    inference_ensemble,
    inference_medium_range,
    schema,
    score_ensemble_outputs,
    weather_events,
)
from earth2mip._channel_stds import channel_stds
from earth2mip.inference_ensemble import run_basic_inference
from earth2mip.networks import get_model, persistence


def run(args):
    return subprocess.check_call(["coverage", "run", *args])  # noqa: S603 S607


def checksum_reduce_precision(arr, digits=3):
    most_significant = max(arr.max(), -arr.min())
    least = most_significant / 10**digits
    arr = (most_significant / least).astype(np.int32)
    checksum = hashlib.md5(arr.data)  # noqa: S324
    return checksum.hexdigest()


class get_data_source:
    def __init__(self, inference):
        self.channel_names = inference.in_channel_names
        self.grid = inference.grid
        arr = np.ones([len(inference.in_channel_names), *inference.grid.shape])
        self.arr = arr
        self.channel_names = inference.out_channel_names

    def __getitem__(self, time) -> np.ndarray:
        return self.arr


def test_inference_ensemble(tmp_path):
    inference = persistence(package=None)
    data_source = get_data_source(inference)
    time = datetime.datetime(2018, 1, 1)
    channel_stds["a"] = 0.0
    channel_stds["b"] = 0.0
    channel_stds["c"] = 0.0
    config = schema.EnsembleRun(
        weather_model="dummy",
        simulation_length=40,
        output_path=tmp_path.as_posix(),
        weather_event=schema.WeatherEvent(
            properties=weather_events.WeatherEventProperties(
                name="test", start_time=time
            ),
            domains=[
                weather_events.Window(
                    name="globe",
                    diagnostics=[
                        weather_events.Diagnostic(
                            type="raw",
                            channels=inference.out_channel_names,
                        )
                    ],
                )
            ],
        ),
    )

    inference_ensemble.run_inference(
        inference, config, data_source=data_source, progress=True
    )

    path = tmp_path / "ensemble_out_0.nc"
    ds = xarray.open_dataset(path.as_posix(), decode_times=False)
    assert ds.time[0].item() == 0
    out = tmp_path / "out"
    score_ensemble_outputs.main(tmp_path.as_posix(), out.as_posix(), score=False)


def test_checksum_reduce_precision(regtest):
    # Test case 1: Basic example
    arr1 = np.array([1.23456, 2.34567, 3.45678])
    arr2 = np.array([1, 2, 3])
    assert checksum_reduce_precision(arr1, digits=1) == checksum_reduce_precision(
        arr2, digits=1
    )

    arr1 = np.array([0.23456, 2.34567, 3.45678])
    arr2 = np.array([1, 2, 3])
    assert checksum_reduce_precision(arr1, digits=1) != checksum_reduce_precision(
        arr2, digits=2
    )

    print(checksum_reduce_precision(arr1), file=regtest)


def test_inference_medium_range(tmpdir, regtest):
    if not torch.cuda.is_available():
        pytest.skip("need gpu")
    inference = persistence(package=None).cuda()
    data_source = get_data_source(inference)
    time = datetime.datetime(2018, 1, 1)
    dt = datetime.timedelta(hours=6)
    times = [time + dt * k for k in range(10)]
    mean = np.zeros((len(inference.out_channel_names), 721, 1440))
    metrics = inference_medium_range.score_deterministic(
        inference, n=5, initial_times=times, time_mean=mean, data_source=data_source
    )
    metrics["acc"].attrs["checksum"] = checksum_reduce_precision(metrics.acc, digits=3)
    metrics["rmse"].attrs["checksum"] = checksum_reduce_precision(
        metrics.rmse, digits=3
    )
    metrics.info(regtest)


@pytest.mark.slow
@pytest.mark.parametrize("url", ["e2mip://fcn", "e2mip://dlwp"])
def test_run_basic_inference(url):
    if url == "e2mip://fcn":
        pytest.xfail(
            "fcn package is broken. See https://github.com/NVIDIA/earth2mip/issues/75"
        )

    time_loop = get_model(url, device="cuda:0")
    data_source = get_data_source(time_loop)
    ds = run_basic_inference(
        time_loop, n=10, data_source=data_source, time=datetime.datetime(2018, 1, 1)
    )
    print(ds)


@pytest.mark.cli
def test_inference_medium_range_cli(tmp_path: pathlib.Path):
    create_hdf5(
        tmp_path,
        2018,
        20,
        grid=earth2mip.grid.equiangular_lat_lon_grid(721, 1440),
        channels=earth2mip.networks.dlwp.CHANNELS,
    )
    output_path = tmp_path / "out"
    run(
        [
            "-m",
            "earth2mip.inference_medium_range",
            "--data",
            tmp_path.as_posix(),
            "-n",
            "5",
            "--start-time",
            "2018-01-02",
            "--end-time",
            "2018-01-02",
            "e2mip://dlwp",
            output_path.as_posix(),
        ]
    )
    series = earth2mip.forecast_metrics_io.read_metrics(output_path.as_posix())
    assert not series.empty


@pytest.mark.cli
def test_lagged_ensemble_cli(tmp_path: pathlib.Path):
    create_hdf5(
        tmp_path,
        2018,
        40,
        grid=earth2mip.grid.equiangular_lat_lon_grid(721, 1440),
        channels=earth2mip.networks.dlwp.CHANNELS,
    )
    output_path = tmp_path / "out"
    run(
        [
            "earth2mip/lagged_ensembles/__main__.py",
            "--data",
            tmp_path.as_posix(),
            "--start-time",
            "2018-01-02",
            "--end-time",
            "2018-01-04",
            "--lags",
            "1",
            "--leads",
            "3",
            "--model",
            "e2mip://dlwp",
            "--output",
            output_path.as_posix(),
        ]
    )
    series = earth2mip.forecast_metrics_io.read_metrics(output_path.as_posix())
    assert not series.empty
