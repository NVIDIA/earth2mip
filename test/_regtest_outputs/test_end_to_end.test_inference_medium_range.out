xarray.Dataset {
dimensions:
	lead_time = 6 ;
	channel = 3 ;
	initial_time = 10 ;

variables:
	timedelta64[ns] lead_time(lead_time) ;
	object channel(channel) ;
	float64 rmse(lead_time, channel) ;
		rmse:checksum = 152afd7bf4dbe26f85032eee0269201a ;
	float64 acc(lead_time, channel) ;
		acc:checksum = 152afd7bf4dbe26f85032eee0269201a ;
	datetime64[ns] initial_time(initial_time) ;
	datetime64[ns] initial_times(initial_time) ;

// global attributes:
}